amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpChannel.java

```diff
@@ -235,56 +234,58 @@ public void receiveBasicConsume(AMQShortString queue, AMQShortString consumerTag
                     closeChannel(ErrorCodes.NOT_FOUND, "No such queue: '" + queue.toString() + "'");
                 } else {
                     PersistentTopic indexTopic = ((PersistentQueue) amqpQueue).getIndexTopic();
-                    try {
-                        AmqpConsumer consumer = subscribe(consumerTag1, queue.toString(), indexTopic,
-                                noAck, exclusive);
-                        if (!nowait) {
-                            MethodRegistry methodRegistry = connection.getMethodRegistry();
-                            AMQMethodBody responseBody = methodRegistry.
-                                    createBasicConsumeOkBody(AMQShortString.
-                                            createAMQShortString(consumer.getConsumerTag()));
-                            connection.writeFrame(responseBody.generateFrame(channelId));
-                        }
-                    } catch (Exception e) {
-                        closeChannel(ErrorCodes.SYNTAX_ERROR, e.getMessage());
-                        log.error("BasicConsume error queue:{} consumerTag:{} ex {}",
-                                queue, consumerTag, e.getMessage());
-                    }
+                    subscribe(consumerTag1, queue.toString(), indexTopic, noAck, exclusive, nowait);
                 }
             }
         });
     }
 
-    private AmqpConsumer subscribe(String consumerTag, String queueName, Topic topic,
-        boolean ack, boolean exclusive) throws ConsumerTagInUseException,
-        InterruptedException, ExecutionException, BrokerServiceException {
+    private void subscribe(String consumerTag, String queueName, Topic topic,
+                           boolean ack, boolean exclusive, boolean nowait){
         if (consumerTag == null) {
             consumerTag = "consumerTag_" + getNextConsumerTag();
         }
+        final String finalConsumerTag = consumerTag;
+
+        CompletableFuture<Void> exceptionFuture = new CompletableFuture<>();
```

Comment:
When we change it to the asynchonized method, the following method will be unpredictable. Because we never know it's done or not. Is it possible to call this method multiple times from the same connection? The first subscriptionFuture calls this method and being blocked but another subscriptionFuture calls this method succecced. So what happens when the first consumer succeed?

Reply:
I think the sync way is the same as the async way for clients, clients will wait to receive the subscribe response from the server-side.